File: my-folder/0206-reverse-linked-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode pointer = null;
    public ListNode reverseList(ListNode head) {
        solve(head);
        return this.pointer;
    }
    
    public ListNode solve (ListNode head) {
        if (head == null || head.next == null) return this.pointer = head;
        solve(head.next).next = head;
        head.next = null;
        return head;
    }
}

File: my-folder/0206-reverse-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode *pointer;
public:

    ListNode* reverseList(ListNode* head) {
        solve(head);
        return pointer;
    }

    ListNode* solve(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            pointer = head;
            return pointer;
        }
        solve(head->next)->next = head;
        head->next = nullptr;
        return head;
    }
};
